### PR TITLE
1 gplazma token exchange plugin

### PR DESCRIPTION
**Pull request by hrhee**
This pull request aims to extend the gPlazma Plugin by adding support for token exchange functionality.

Changes Introduced:

- Added Token Exchange Functionality: Implemented the token exchange feature within the "oidc-te" plugin. This includes the ability to exchange a given token for another token.
- Added following configuration parameters in `skel/share/defaults/gplamza.properties`:
    - gplazma.oidc-te.url
    - gplazma.oidc-te.client-id
    - plazma.oidc-te.client-secret
    - gplazma.oidc-te.grant-type
    - gplazma.oidc-te.subject-issuer
    - gplazma.oidc-te.subject-token-type
    - gplazma.oidc-te.audience
- Unit Tests: Implemented most simple unit tests of the token exchange functionality. To be extended to cover various scenarios, including valid and invalid token exchange requests, error handling, and edge cases.

### CODE REVIEW

**Comment by kofemann** [comment]
ok to test

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    private final static String CLIENT_SECRET = "gplazma.oidc-te.client-secret";
+    private final static String GRANT_TYPE = "gplazma.oidc-te.grant-type";
+    private final static String SUBJECT_ISSUER = "gplazma.oidc-te.subject-issuer";
+    private final static String SUBJECT_TOKEN_TYPE = "gplazma.oidc-te.subject-token-type";
+    private final static String AUDIENCE = "gplazma.oidc-te.audience";
+
+    private final CloseableHttpClient client;
+
+    public TokenExchange () {
+        this.client = HttpClients.createDefault();
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:47]: maybe 

```java
this(HttpClients.createDefault());
```

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+            }
+        }
+
+        String token = credential.getToken();
+        LOG.debug("Found bearer token: {}", token);
+
+        checkAuthentication(token != null, "No bearer token in the credentials");
+
+        String exchangedToken = null;
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:75]: the variable should be defined within try-catch block

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        } catch ( IOException | URISyntaxException e ) {
+            throw new AuthenticationException("Unable to process token: " + e.getMessage());
+        }
+
+    }
+
+
+    @VisibleForTesting
+    public String tokenExchange(String token) throws IOException, URISyntaxException {
+        String postBody = "client_id=" + CLIENT_ID
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:91]: Using the property names makes no sense here...

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            privateCredentials.add(new BearerTokenCredential(exchangedToken));
+
+        } catch ( IOException | URISyntaxException e ) {
+            throw new AuthenticationException("Unable to process token: " + e.getMessage());
+        }
+
+    }
+
+
+    @VisibleForTesting
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:89]: this is a public method. The annotation makes sense if you make limit scope to the package (just drop the `public` keyword)

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    public String tokenExchange(String token) throws IOException, URISyntaxException {
+        String postBody = "client_id=" + CLIENT_ID
+                + "&client_secret=" + CLIENT_SECRET
+                + "&grant_type=" + GRANT_TYPE
+                + "&subject_token=" + token
+                + "&subject_issuer=" + SUBJECT_ISSUER
+                + "&subject_token_type=" + SUBJECT_TOKEN_TYPE
+                + "&audience=" + AUDIENCE;
+    
+        URI uri = new URIBuilder(TOKEN_EXCHANGE_URL).build();
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:99]: same here....

File: modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
+/* dCache - http://www.dcache.org/
+ *
+ * Copyright (C) 2022 Deutsches Elektronen-Synchrotron
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java:3]: 2024?

File: modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+public class TokenExchangeTest {
+
+    private TokenExchange plugin; 
+    private CloseableHttpClient httpClient;
+
+    @Test(expected=IOException.class)
+    public void shouldFailWithInvalidToken() throws Exception {
+        plugin = new TokenExchange();
+        // plugin = aPlugin().build();
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java:44]: please remove dead code

File: modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+        plugin.tokenExchange("InvalidToken"); 
+    }
+
+    @Test
+    public void tokenExchangeTest() throws Exception {
+
+        given(aPlugin().withAuthorizationServer(aAuthorizationServer().thatExchanges()));
+        // plugin = new TokenExchange();
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java:53]: ☝🏼

File: modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        String result = plugin.tokenExchange(token);
+
+        // System.out.println("result: " + result);
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java:58]: ☝🏼

File: modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            CloseableHttpResponse mockResponse = Mockito.mock(CloseableHttpResponse.class);
+            String jsonContent = "{\"access_token\": \"valid.access.token\"}";
+            try {
+                HttpEntity entity = new StringEntity(jsonContent);
+                Mockito.when(mockResponse.getEntity()).thenReturn(entity);
+            } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException("Impossible exception caught", e);
+            }
+        
+            try {
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/test/java/org/dcache/gplazma/tokenx/TokenExchangeTest.java:92]: you can merge this two try-catch blocks

File: packages/system-test/src/main/bin/populate [hunk 11]
```diff
@@ -138,7 +138,12 @@ EOF
     host_cert_hash=$(openssl x509 -in $ca_cert -noout -subject_hash)
     cp "$ca_cert" "$trust_store/$host_cert_hash.0"
 
+# <<<<<<< HEAD
```
Comment (kofemann) [inline packages/system-test/src/main/bin/populate:141]: please remove dead code

File: packages/system-test/src/main/bin/populate [hunk 12]
```diff
@@ -138,7 +138,12 @@ EOF
...
+#     # old_dn="$(openssl x509 -in $ca_cert -noout -subject -nameopt compat | sed 's:subject= *\+/\?:/:;s:, :/:g')"
```
Comment (kofemann) [inline packages/system-test/src/main/bin/populate:142]: ☝🏼

File: packages/system-test/src/main/bin/populate [hunk 13]
```diff
@@ -283,7 +288,13 @@ add_dn_from_file() { # $1 - X.509 pem certificate
 	return
     fi
 
+# <<<<<<< HEAD
+#     # dn=$(openssl x509 -in "$1" -subject -noout -nameopt compat | sed 's:subject= *\+/\?:/:;s:, :/:g')
+#     # dn=$(openssl x509 -in "$1" -subject -noout -nameopt compat | sed -E 's:subject= */:?/:;s:details= */:?/:;s:priority= */:?/:' )
+#     dn=$(openssl x509 -in "$1" -subject -noout -nameopt compat | sed -E 's:subject= */?\?:/:;s: *,:/:g' )
+# =======
```
Comment (kofemann) [inline packages/system-test/src/main/bin/populate:295]: ☝🏼

File: skel/share/defaults/gplazma.properties [hunk 14]
```diff
@@ -519,6 +519,18 @@ gplazma.oidc.access-token-cache.expire.unit = SECONDS
 #
 gplazma.oidc.audience-targets =
 
+# ---- OpenID Connect Token Exchange plugin
+#
+
+gplazma.oidc-te.url = https://keycloak.desy.de/auth/realms/production/protocol/openid-connect/token
```
Comment (kofemann) [inline skel/share/defaults/gplazma.properties:525]: no real host names. People might forget to change it and will start accessing desy keycloak

File: skel/share/defaults/gplazma.properties [hunk 15]
```diff
@@ -519,6 +519,18 @@ gplazma.oidc.access-token-cache.expire.unit = SECONDS
...
+gplazma.oidc-te.client-id = token-exchange 
+gplazma.oidc-te.client-secret = xxx
+gplazma.oidc-te.grant-type = urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange
```
Comment (kofemann) [inline skel/share/defaults/gplazma.properties:528]: this kind of format is not really user-friendly. You should take a regular string and encode it in the plugin.

**Review by kofemann — CHANGES_REQUESTED** [review]
Well Done! 🎉

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 1 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java]: Hi @kofemann, very sorry for the long delay, but I would like to pick up this long overdue PR.

 I have adopted your suggestion and modified the constructor to use `this`.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 2 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java]: Thanks, I have moved the variable within the try-catch block.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 3 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java]: yes, this makes no sense here.

I have modified the code to use the the `Properties` class to read the parameters.
But I am not certain that the values are taken from `skel/share/defaults/gplazma.properties`

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 4 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:109]: yes, I have removed the keyword `public`.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 5 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java]: (see above for other parameters)

File: packages/system-test/src/main/bin/populate (hunk 11 above)
Comment (hrhee) [reply packages/system-test/src/main/bin/populate]: sorry, dead code is removed.

File: packages/system-test/src/main/bin/populate (hunk 12 above)
Comment (hrhee) [reply packages/system-test/src/main/bin/populate]: dead code removed.

File: packages/system-test/src/main/bin/populate (hunk 13 above)
Comment (hrhee) [reply packages/system-test/src/main/bin/populate]: dead code removed.

File: skel/share/defaults/gplazma.properties (hunk 14 above)
Comment (hrhee) [reply skel/share/defaults/gplazma.properties]: removed the real host name. Instead the value for `gplamza.oidc.provider` is reused for the token exchange endpoint.

File: skel/share/defaults/gplazma.properties (hunk 15 above)
Comment (hrhee) [reply skel/share/defaults/gplazma.properties]: changed the parameters to a more user-friendly format.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+        String token = credential.getToken();
+        LOG.debug("Found bearer token: {}", token);
+
+        checkAuthentication(token != null, "No bearer token in the credentials");
+
+
+        try {
+            String exchangedToken = null;
+            exchangedToken = tokenExchange(token);
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:98]: `String exchangedToken = tokenExchange(token);`

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+        for (Object c : privateCredentials) {
+            if (c instanceof BearerTokenCredential) {
+
+                credential = (BearerTokenCredential) c;
+                checkAuthentication(credential == null, "Multiple bearer token credentials");
+
+            }
+        }
+
+        String token = credential.getToken();
```
[the remark this answers is not in the document]
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:94]: thanks for the remark! I modified the code so that a AuthenticationException is thrown if no BearerTokenCredential is found.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 16 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java]: modified as you suggested.

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    public void authenticate(Set<Object> publicCredentials, Set<Object> privateCredentials,
+        Set<Principal> identifiedPrincipals, Set<Restriction> restrictions)
+        throws AuthenticationException {
+
+        BearerTokenCredential credential = null;
+        for (Object c : privateCredentials) {
+            if (c instanceof BearerTokenCredential) {
+
+                credential = (BearerTokenCredential) c;
+                checkAuthentication(credential == null, "Multiple bearer token credentials");
```
Comment (kofemann) [inline modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:85]: the `credential` is never null, as you just have assigned to it a new value. The check must be performed before the assignment (swap the lines).

**Review by kofemann — CHANGES_REQUESTED** [review]
A single correction and the code is ready to be merged. Thanks!

**Review by kofemann — APPROVED** [review]
Thanks!

File: modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java (hunk 18 above)
Comment (hrhee) [reply modules/gplazma2-oidc-te/src/main/java/org/dcache/gplazma/tokenx/TokenExchange.java:84]: yes, I have swaped the lines. The null check is done before the assignment.

**Comment by hrhee** [comment]
Hi @kofemann, this PR is not anymore in draft mode and is ready for further review or merge. Thanks a lot!
